opentracing/span.py

```diff
@@ -228,3 +223,58 @@ def log(self, **kwargs):
         if 'timestamp' in kwargs:
             timestamp = kwargs['timestamp']
         return self.log_kv(key_values, timestamp)
+
+
+class ContinuationMixin(object):
+    """
+    Continuation mixin that provides additional primitives to a `Span` class,
+    so that it can be deactivated and resumed later during in-process
+    propagation.
+
+    If the application needs to defer work that should be part of the same
+    `Span`, the `ContinuationMixin` provides a `capture` method that increments
+    a counter to the span so that it can be deactivated without finishing. In
+    another asynchronous executor and/or thread, it can be resumed.
+    """
+    def capture(self):
+        """
+        Capture the `Span` tracing session as well as any 3rd-party execution
+        context of interest. After calling this method, the `Span` may be used
+        later in a closure or callback function where it may be resumed and
+        reactivated using the tracer `active_span_source.make_active()`.
+
+        IMPORTANT: the caller MUST activate and then deactivate the `Span`
+        otherwise it will never automatically finish. That is, calling
+        `capture()` increments a refcount that must be decremented somewhere
+        else.
+        """
+        pass
+
+    def deactivate(self):
+        """
+        Mark the end of the active period for the current `Span` in this
+        asynchronous executor and/or thread. When the last refcount has been
+        decremented through `deactivate()`, it is automatically finished.
+
+        NOTE: Calling `deactivate()` more than once on a single `Span` instance
+        leads to undefined behavior.
+        """
+        self.finish()
+
+
+class Span(ContinuationMixin, BaseSpan):
```

Comment:
@palazzem apologies if I'm missing something here – I'm catching up on a lot of stuff and definitely might just be misunderstanding something.

That said, I'm a little confused about this... won't a real/production implementation of a Continuation need to store the `capture()`-time active `Span`? If so, how would it work for multiple Continuations to actually be the same object instance in memory? (I.e., won't the active `Span` reference "collide"?)

To make this more concrete, what happens if I call `span.capture()` *on the same `Span` instance* from multiple threads? I.e., what have I "captured", and where?

Again, apologies if I'm missing something. If it's easier to understand, I'm basically talking about the `toRestore` field in https://github.com/opentracing/opentracing-java/blob/master/opentracing-util/src/main/java/io/opentracing/util/ThreadLocalActiveSpan.java#L35 ... since there's a `many:1` relationship between `ActiveSpan:Span` in the Java world, it's important that the ActiveSpan has its own storage.